File: Projects/Ex36Thread/app/src/main/java/com/hyun/ex36thread/MainActivity.java
package com.hyun.ex36thread;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.Log;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    int num=0;
    TextView tv;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tv= findViewById(R.id.tv);

        findViewById(R.id.btn).setOnClickListener(v -> {
            // 오래 걸리는 작업 -- 별도 Thread를 만들지 않았을때..
            for (int i=0; i<20; i++) {
                num++;
                tv.setText(num+"");

                //개발자가 화면 뷰를 이용하지 않고 특정값을 확인해보고 싶다면..
                //또는 코드의 실행에 대한 기록을 남기고 싶다면..
                //이 때 사용하는 것이 로그 Log..
                Log.d("aaa", "num: "+num);

                //0.5초 대기
                try {
                    Thread.sleep(500);
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
            }
        });

        findViewById(R.id.btn2).setOnClickListener(v -> {
            //오래 걸리는 작업(ex. network, DB작업...)
            //을 하는 직원(MyThread)객체 생성 및 실행
            MyThread t=new MyThread();
            //t.run(); //run()을 직접 호출하면.. Main Thread가 작업해버림
            t.start(); //자동으로 run() 실행됨

        });

    }//onCreate method..

    //inner class
    class MyThread extends  Thread{
        //이 특별한 메소드 안에 있는 코드만 이 스레드가 실행함
        @Override
        public void run() {
            for( int i=0; i<20; i++) {
                num++;

                //별도의 Thread는 UI 작업을 수행할 수 없음..
                //tv.setText(num+"");
                //Main Thread 에게 UI 작업을 요청해야 함.
                //요청 방법1. Handler 객체를 이용
                //handler.sendEmptyMessage(0);

                //요청 방법2. 별도의 직원에게 UI 작업 위임장을 부여하는 방법
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        //이 영역안에서는 Ui작업 가능함
                        tv.setText(num+"");
                    }
                });


                try {
                    Thread.sleep(500);
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
            }
        }
    }

    // MainActivity class 의 멤버변수
    Handler handler= new Handler(Looper.getMainLooper()){
        @Override
        public void handleMessage(@NonNull Message msg) {
            //super.handleMessage(msg);
            // 이 영역안에서는 UI 작업이 가능함
            tv.setText(num+"");
        }
    };

}//MainActivity class

File: Projects/Ex52PhotoPickMultiple/app/src/main/java/com/hyun/ex52photopickmultiple/MainActivity.kt
package com.hyun.ex52photopickmultiple

import android.content.Intent
import android.net.Uri
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.PickVisualMediaRequest
import androidx.activity.result.contract.ActivityResultContracts
import androidx.viewpager2.widget.ViewPager2
import com.tbuonomo.viewpagerdotsindicator.WormDotsIndicator

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        findViewById<Button>(R.id.btn1).setOnClickListener{ clickBtn1() }
        findViewById<Button>(R.id.btn2).setOnClickListener{ clickBtn2() }
    }

    val tv: TextView by lazy { findViewById(R.id.tv) }

    // 선택된 파일들의 uri 를 저장하는 리스트
    val imgs: MutableList<Uri?> = mutableListOf()
    val pager: ViewPager2 by lazy { findViewById(R.id.pager) }
    val dotsIndicator: WormDotsIndicator by lazy { findViewById(R.id.dots_indicator) }

    private fun clickBtn1(){
        val intent: Intent = Intent(Intent.ACTION_OPEN_DOCUMENT).setType("image/*").putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true)
        resultLauncher.launch(intent)

    }

    //결과 받아주는 대행사 객체 등록
    val resultLauncher: ActivityResultLauncher<Intent> = registerForActivityResult(ActivityResultContracts.StartActivityForResult()){
            result ->

        if(result.resultCode == RESULT_CANCELED ){
            tv.text="선택을 하지 않았습니다."
        }else{
            imgs.clear()

            // 1개만 선택했을때는.. uri data로 전달받음
            if(result.data?.data != null){
                imgs.add(result.data?.data)

            }else{// 2개 이상일때틑 ClipData 라는 것으로 여러 파일들의 정보를 받음
                val cnt:Int = result.data?.clipData?.itemCount!!
                for( i in 0 until cnt){
                    imgs.add(  result.data?.clipData?.getItemAt(i)?.uri )
                }
            }

            tv.text= "선택한 이미지 개수 : ${imgs.size}"
            pager.adapter = MyAdapter(this, imgs)
            dotsIndicator.attachTo(pager)


        }

    }


    private fun clickBtn2(){
        // JetPack 에서 새로 추가된 기능
        pickMultipleLauncher.launch(PickVisualMediaRequest(ActivityResultContracts.PickVisualMedia.ImageOnly))
    }

    val pickMultipleLauncher: ActivityResultLauncher<PickVisualMediaRequest> = registerForActivityResult(ActivityResultContracts.PickMultipleVisualMedia()){
            uris: List<Uri> ->

        imgs.clear()

        for( uri in uris) imgs.add(uri)

        tv.text= "선택한 이미지 개수 : ${imgs.size}"
        pager.adapter= MyAdapter(this, imgs)
        dotsIndicator.attachTo(pager)

    }

}













